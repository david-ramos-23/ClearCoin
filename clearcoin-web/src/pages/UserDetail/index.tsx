import React, {useEffect, useMemo, useState} from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import styled from 'styled-components';
import { getUser, transfer, resetErrorMsg } from '../../store/user/actions';
import DataTable from 'react-data-table-component';
import Modal from '../../components/Modal';
import Card from '../../components/Card';
import { Transfer, Wallet } from '../../types/types';
import FilterTable from '../../components/FilterTable'

const Container = styled.div`
    width: 80%;
    max-width: 1100px;
`;

const UserInfo = styled.div`
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: flex-end;

`;

const BalanceWrapper = styled.div`
    display: flex;
    justify-content: center;
    margin-top: 2rem;
`;
const Balance = styled.h2`
    text-align: center;
    border-radius: .75rem;
    padding: 1rem;
    background-color: rgba(0,0,0,0.2);
`;

const getBalance = (wallets: any) => {
    return wallets.reduce((acc: number, item: any) => {return acc + item.amount}, 0);
}

const UserDetail = () => {
    const dispatch = useDispatch();
    const params: any = useParams();
    const [isModalVisible, setIsModalVisible] = useState(false);
    const [selectedWallet, setSelectedWallet] = useState<Wallet>({} as Wallet);
    const detailId = params.detailId ? parseInt(params.detailId, 10) : 0;
    const { userDetail, hasDataChanged, errorMsg } = useSelector((state: any) => state.user);

    const onSubmitForm = (values: Transfer) => {
        dispatch(transfer(values));
    };

    const columns = useMemo(() => [
    {
        name: 'Wallet ID',
        selector: 'id',
        sortable: true,
        ignoreRowClick: true
    },
    {
        name: 'Amount',
        selector: 'amount',
        sortable: true,
        
    }], []);

    useEffect(() => {
        setIsModalVisible(false);
        dispatch(getUser(detailId));
    }, [dispatch, detailId, hasDataChanged]);
    

    const [filterText, setFilterText] = React.useState('');
    const [resetPaginationToggle, setResetPaginationToggle] = React.useState(false);
    const filteredItems = userDetail.wallets.filter((wallets: Wallet) => wallets.id && wallets.id.toLowerCase().includes(filterText.toLowerCase()));

    const subHeaderComponentMemo = React.useMemo(() => {
    const handleClear = () => {
        if (filterText) {
            setResetPaginationToggle(!resetPaginationToggle);
            setFilterText('');
        }
    };

    return <FilterTable onFilter={(e: any) => setFilterText(e.target.value)} onClear={handleClear} filterText={filterText} filterBy="Filter By Wallet ID" />;
    }, [filterText, resetPaginationToggle]);

    const onRowClicked = (row: any) => {
        setSelectedWallet(row);
        setIsModalVisible(true);
    }

    const onModalClose = () => {
        setIsModalVisible(false);
        dispatch(resetErrorMsg());
    }

    useEffect(() => {
        setIsModalVisible(false);
        dispatch(getUser(detailId));
    }, [dispatch, detailId, hasDataChanged]);

    return (
        <Container>
            <Modal visible={isModalVisible} onClose={onModalClose}>
                <Card wallet={selectedWallet} onSubmit={onSubmitForm} errorMsg={errorMsg}/>
            </Modal>
            <UserInfo>
                <div>
                    <h3>First name</h3>
                    <span>{userDetail.firstName}</span>
                </div>

                <div>
                    <h3>Last name</h3>
                    <span>{userDetail.lastName}</span>
                </div>
            
                <div>
                    <h3>Email</h3>
                    <span>{userDetail.email}</span>
                </div>
            </UserInfo>

            <BalanceWrapper>
                <Balance>Balance {`${getBalance(userDetail.wallets)}â‚¬`}</Balance>
            </BalanceWrapper>
            
            <DataTable
                columns={columns}
                data={filteredItems}
                onRowClicked={onRowClicked}
                paginationResetDefaultPage={resetPaginationToggle} // optionally, a hook to reset pagination to page 1
                subHeader
                subHeaderComponent={subHeaderComponentMemo}
                subHeaderWrap={false}
                persistTableHead
                pointerOnHover
                highlightOnHover	
            />
        </Container>
    );
}

export default UserDetail;
