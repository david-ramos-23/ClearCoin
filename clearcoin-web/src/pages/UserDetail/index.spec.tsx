import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { render, screen } from '@testing-library/react';
import UserDetail from './index';
const { user } = require('../../mock/userMock');

let mockUserSelector: any;

jest.mock('react-redux', () => ({
	useSelector: () => mockUserSelector,
    useDispatch: () => {
		return jest.fn();
	},
    useParams: () => {
		return "1";
	},
}));

describe('User', () => {
	test('Render User Page', () => {
		const history = createMemoryHistory();
		mockUserSelector = {userDetail:user, hasDataChanged: false, errorMsg: undefined};

		render(
				<Router history={ history }>
						<UserDetail/>
				</Router>);

		expect(screen.getByText('David')).toBeInTheDocument();
		expect(screen.getByText('Ramos')).toBeInTheDocument();
		expect(screen.getByText('andara14@gmail.com')).toBeInTheDocument();
		expect(screen.getByText('Balance 2326.94â‚¬')).toBeInTheDocument();
	});

	test('snapshot is ok', () => {
		const history = createMemoryHistory();
		mockUserSelector = {userDetail: user, hasDataChanged: false, errorMsg: undefined};

		const { container } = render(
			<Router history={ history }>
					<UserDetail/>
			</Router>);

		expect(container).toMatchSnapshot();
	});
});
