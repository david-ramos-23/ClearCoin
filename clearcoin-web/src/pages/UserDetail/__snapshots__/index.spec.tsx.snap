// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User snapshot is ok 1`] = `
.c15 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
  max-width: 100%;
  color: rgba(0,0,0,0.87);
  background-color: #FFFFFF;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  width: 100%;
  background-color: #FFFFFF;
  min-height: 56px;
  border-bottom-width: 1px;
  border-bottom-color: rgba(0,0,0,.12);
  border-bottom-style: solid;
}

.c18 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  font-size: 12px;
  font-weight: 500;
  color: rgba(0,0,0,0.87);
  padding-left: 16px;
  padding-right: 16px;
}

.c24 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  padding-left: 16px;
  padding-right: 16px;
  word-break: break-word;
}

.c19 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  max-width: 100%;
  min-width: 100px;
}

.c25 {
  font-weight: 400;
}

.c25 div:first-child {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c23 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: 100%;
  box-sizing: border-box;
  font-size: 13px;
  color: rgba(0,0,0,0.87);
  background-color: #FFFFFF;
  min-height: 48px;
}

.c23:not(:last-of-type) {
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-bottom-color: rgba(0,0,0,.12);
}

.c23:hover {
  color: rgba(0,0,0,0.87);
  background-color: #EEEEEE;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  border-bottom-color: #FFFFFF;
  outline-style: solid;
  outline-width: 1px;
  outline-color: #FFFFFF;
}

.c23:hover {
  cursor: pointer;
}

.c21 {
  padding: 2px;
  color: inherit;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  opacity: 0;
}

.c20 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  line-height: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c20:focus {
  outline: none;
  color: rgba(0,0,0,.54);
}

.c20:hover {
  color: rgba(0,0,0,.54);
}

.c20 span.__rdt_custom_sort_icon__ i,
.c20 span.__rdt_custom_sort_icon__ svg {
  opacity: 0;
  color: inherit;
  font-size: 18px !important;
  height: 18px !important;
  width: 18px !important;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
}

.c20 span.__rdt_custom_sort_icon__.asc i,
.c20 span.__rdt_custom_sort_icon__.asc svg {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c20:hover {
  cursor: pointer;
}

.c20:hover span,
.c20:hover span.__rdt_custom_sort_icon__ * {
  opacity: 1;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  height: 100%;
  font-weight: 400;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c7 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  box-sizing: inherit;
  z-index: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #e3f2fd;
  font-size: 18px;
  font-weight: 400;
  color: rgba(0,0,0,0.87);
  padding-left: 16px;
  padding-right: 8px;
  -webkit-transform: translate3d(0,-100%,0);
  -ms-transform: translate3d(0,-100%,0);
  transform: translate3d(0,-100%,0);
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
  -webkit-transition-timing-function: cubic-bezier(0,0,0.2,1);
  transition-timing-function: cubic-bezier(0,0,0.2,1);
  will-change: transform;
}

.c4 {
  position: relative;
  box-sizing: border-box;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  font-size: 22px;
  color: rgba(0,0,0,0.87);
  background-color: #FFFFFF;
  min-height: 56px;
  padding-left: 16px;
  padding-right: 8px;
}

.c5 {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font-weight: 400;
}

.c6 {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c6 > * {
  margin-left: 5px;
}

.c10 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 4px 16px 4px 24px;
  width: 100%;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  background-color: #FFFFFF;
  min-height: 52px;
}

.c22 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c13 {
  position: relative;
  width: 100%;
  border-radius: inherit;
  overflow-x: auto;
  overflow-y: hidden;
  min-height: 0;
}

.c14 {
  position: relative;
  width: 100%;
  display: table;
}

.c11 {
  height: 32px;
  width: 200px;
  border-radius: 3px;
  border-top-left-radius: 5px;
  border-bottom-left-radius: 5px;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border: 1px solid #e5e5e5;
  padding: 0 32px 0 16px;
}

.c11:hover {
  border: 1px solid #b2fce4;
  cursor: pointer;
}

.c11:focus {
  border: 1px solid #b2fce4;
  cursor: pointer;
}

.c12 {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  border-top-right-radius: 5px;
  border-bottom-right-radius: 5px;
  height: 34px;
  width: 32px;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: none;
  color: black;
  padding: 8px 32px;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 16px;
  background-color: #b2fce4;
  -webkit-transition: all .2s ease-in;
  transition: all .2s ease-in;
}

.c12:hover {
  background-color: #828282;
  color: #fff;
  cursor: pointer;
}

.c0 {
  width: 80%;
  max-width: 1100px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 2rem;
}

.c3 {
  text-align: center;
  border-radius: .75rem;
  padding: 1rem;
  background-color: rgba(0,0,0,0.2);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div>
        <h3>
          First name
        </h3>
        <span>
          David
        </span>
      </div>
      <div>
        <h3>
          Last name
        </h3>
        <span>
          Ramos
        </span>
      </div>
      <div>
        <h3>
          Email
        </h3>
        <span>
          andara14@gmail.com
        </span>
      </div>
    </div>
    <div
      class="c2"
    >
      <h2
        class="c3"
      >
        Balance 
        2326.94€
      </h2>
    </div>
    <div
      aria-level="1"
      class="c4 rdt_TableHeader"
      role="heading"
    >
      <div
        class="c5"
      />
      <div
        class="c6"
      />
      <div
        class="c7"
      >
        <div
          class="c8"
        />
        <div
          class="c9"
        />
      </div>
    </div>
    <header
      class="c10"
    >
      <input
        aria-label="Search Input"
        class="c11"
        id="search"
        placeholder="Filter By Wallet ID"
        type="text"
        value=""
      />
      <button
        class="c12"
        type="button"
      >
        X
      </button>
    </header>
    <div
      class="c13"
    >
      <div
        class="c14"
      >
        <div
          class="c15 rdt_Table"
          role="table"
        >
          <div
            class="c16 rdt_TableHead"
            role="rowgroup"
          >
            <div
              class="c17 rdt_TableHeadRow"
              role="row"
            >
              <div
                class="c18 c19 rdt_TableCol"
              >
                <div
                  class="c20 rdt_TableCol_Sortable"
                  id="column-id"
                  role="columnheader"
                  tabindex="0"
                >
                  <div>
                    Wallet ID
                  </div>
                  <span
                    class="c21"
                  >
                    ▲
                  </span>
                </div>
              </div>
              <div
                class="c18 c19 rdt_TableCol"
              >
                <div
                  class="c20 rdt_TableCol_Sortable"
                  id="column-amount"
                  role="columnheader"
                  tabindex="0"
                >
                  <div>
                    Amount
                  </div>
                  <span
                    class="c21"
                  >
                    ▲
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="c22 rdt_TableBody"
            offset="250px"
            role="rowgroup"
          >
            <div
              class="c23 rdt_TableRow"
              id="row-ad55e955-1dde-40b1-a308-a6ed40cd13f0"
              role="row"
            >
              <div
                class="c24 c19 c25 rdt_TableCell"
                id="cell--XB_BYZ7zh-ad55e955-1dde-40b1-a308-a6ed40cd13f0"
                role="gridcell"
              >
                <div>
                  ad55e955-1dde-40b1-a308-a6ed40cd13f0
                </div>
              </div>
              <div
                class="c24 c19 c25 rdt_TableCell"
                data-tag="allowRowEvents"
                id="cell-udjr8Mxzh2-ad55e955-1dde-40b1-a308-a6ed40cd13f0"
                role="gridcell"
              >
                <div
                  data-tag="allowRowEvents"
                >
                  2303.94
                </div>
              </div>
            </div>
            <div
              class="c23 rdt_TableRow"
              id="row-1161faf0-4e4f-41a9-aeef-6446918fdcb6"
              role="row"
            >
              <div
                class="c24 c19 c25 rdt_TableCell"
                id="cell--XB_BYZ7zh-1161faf0-4e4f-41a9-aeef-6446918fdcb6"
                role="gridcell"
              >
                <div>
                  1161faf0-4e4f-41a9-aeef-6446918fdcb6
                </div>
              </div>
              <div
                class="c24 c19 c25 rdt_TableCell"
                data-tag="allowRowEvents"
                id="cell-udjr8Mxzh2-1161faf0-4e4f-41a9-aeef-6446918fdcb6"
                role="gridcell"
              >
                <div
                  data-tag="allowRowEvents"
                >
                  23
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
